### PR TITLE
✨Feat(cdk): make sure file-transfer is handled when _concurrency_level is set and new get_file method for connectors.

### PR DESCRIPTION
**Pull request by aldogonzalez8**
## What
For file transfer changes, there is a scenario we were missing, which is when file_base_source.py encounters `_concurrency_level` set in the source connector e.g. source-s3, so this PR is intended to amend that.

UPDATE:
I was not having good results for S3 connector with file pointer + read approach so I'm moving to allow the connector to provide the method to read/download/get the file like here [Feat(file-transfer): add file-transfer feature to s3](https://github.com/airbytehq/airbyte/pull/48346/files#diff-b5e8d7cf7e79880dd2e77408410825335670decbc45f4960d2cf07ac323bae86) 

## How
- Add the missing check to make the default stream if the use_file_stream flag is on and related code.

- Allow the connector to provide a mode to get the file

## Review guide

Fie-transfer changes for _concurrency_level set: 

1. `airbyte-cdk/python/airbyte_cdk/sources/file_based/file_based_source.py`: add use_file_transfer flag when _concurrency_level is set.
2. `airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partitions/record.py`: Record class acts as an envelop for [the record message](https://github.com/airbytehq/airbyte/blob/9a51bce6f63e099a49f0f24dee9fd9dedf06ef29/airbyte-cdk/python/airbyte_cdk/models/airbyte_protocol.py#L80) so I think is correct to add `is_file_transfer_message` flag that will match the [stream_data_to_airbyte_message ](https://github.com/airbytehq/airbyte/blob/9a51bce6f63e099a49f0f24dee9fd9dedf06ef29/airbyte-cdk/python/airbyte_cdk/sources/utils/record_helper.py#L15-L21) signature to create the correct message.
3. `airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/concurrent_read_processor.py`: update params passed to `stream_data_to_airbyte_message`.
4. `airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/concurrent/adapters.py`: initialize correct data for Record when file_transfer is set for the stream.

File-transfer changes for stream-reader:

1. `airbyte-cdk/python/airbyte_cdk/sources/file_based/file_types/blob_transfer.py`: Now we expect the stream reader to prove a `get_file()` method where the connector can implement "best" logic to get/read/download files instead of expect the cdk to do it using a file pointer.
2. `airbyte-cdk/python/airbyte_cdk/sources/file_based/file_based_stream_reader.py`: New abstract method for get_file that connector should implement for file transfer.

## User Impact
These changes are required in order to enable File Transfer for source-s3 in this [other PR](https://github.com/airbytehq/airbyte/pull/48346).

## Can this PR be safely reverted and rolled back?

- [ ] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partitions/record.py [hunk 1]
```diff
@@ -13,9 +13,10 @@ class Record:
     Represents a record read from a stream.
     """
 
-    def __init__(self, data: Mapping[str, Any], partition: "Partition"):
+    def __init__(self, data: Mapping[str, Any], partition: "Partition", is_file_transfer_message: bool = False):
         self.data = data
```
Comment (aldogonzalez8) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partitions/record.py:17]: Record class acts as an envelop for [the record message](https://github.com/airbytehq/airbyte/blob/9a51bce6f63e099a49f0f24dee9fd9dedf06ef29/airbyte-cdk/python/airbyte_cdk/models/airbyte_protocol.py#L80) so I think is correct to add is_file_transfer_message flag that will match the [stream_data_to_airbyte_message ](https://github.com/airbytehq/airbyte/blob/9a51bce6f63e099a49f0f24dee9fd9dedf06ef29/airbyte-cdk/python/airbyte_cdk/sources/utils/record_helper.py#L15-L21)signature to create the correct message.

That said, I'm not totally opposed to the idea of creating a different class like FileRecord with a file attribute if anybody feels that is not the right approach. However, following the concept that the record and the record message are not the same, just this class with the flag makes sense, IMO.

File: airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/concurrent/adapters.py [hunk 2]
```diff
@@ -231,7 +231,12 @@ def read(self) -> Iterable[Record]:
                     yield Record(data_to_return, self)
                 elif isinstance(record_data, AirbyteMessage) and record_data.type == Type.RECORD and record_data.record is not None:
                     # `AirbyteMessage`s of type `Record` should also be yielded so they are enqueued
-                    yield Record(record_data.record.data, self)
+                    # If stream is flagged for file_transfer the record should data in file key
+                    record_message_data = record_data.record.file if self._use_file_transfer() else record_data.record.data
+                    if record_message_data is None:
+                        raise ExceptionWithDisplayMessage("A record without data was found")
```
Comment (aldogonzalez8) [inline airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/concurrent/adapters.py:237]: I think this shouldn't happen unless somebody knows a scenario for empty record data "{}", I'm trying to catch any error if the wrong data field is selected/set.

File: airbyte-cdk/python/airbyte_cdk/sources/file_based/file_types/blob_transfer.py [hunk 3]
```diff
@@ -41,19 +24,12 @@ def write_streams(
...
         logger: logging.Logger,
     ) -> Iterable[Dict[str, Any]]:
-        file_no = 0
         try:
-            data_generator = self._file_reader.read_data(config, file, stream_reader, logger, self.file_read_mode)
-            local_writer = LocalFileTransferClient()
-            for file_opened, file_size in data_generator:
-                yield local_writer.write(file.uri, file_opened, file_size, logger)
-                file_no += 1
+            yield stream_reader.get_file(file=file, local_directory=self._local_directory, logger=logger)
```
Comment (aldogonzalez8) [inline airbyte-cdk/python/airbyte_cdk/sources/file_based/file_types/blob_transfer.py:28]: Implementation sample for source-s3 [here](https://github.com/airbytehq/airbyte/pull/48346/files#diff-b5e8d7cf7e79880dd2e77408410825335670decbc45f4960d2cf07ac323bae86).

Comment (aldogonzalez8) [reply airbyte-cdk/python/airbyte_cdk/sources/file_based/file_types/blob_transfer.py:28]: Manual testing in [this workspace](https://cloud.airbyte.com/workspaces/166f3e04-9021-4f63-a963-a43abb548cec/connections/d7e111cf-1ddc-41bc-8b1a-34bf435c3f2f/timeline) seems be working fine for sync > 5GB :)

File: airbyte-cdk/python/airbyte_cdk/sources/file_based/file_types/blob_transfer.py [hunk 4]
```diff
@@ -41,19 +24,12 @@ def write_streams(
...
-            logger.error("An error has occurred while writing file: %s", str(ex))
+            logger.error("An error has occurred while getting file: %s", str(ex))
             raise ex
-        finally:
-            data_generator.close()
 
     @property
     def file_read_mode(self) -> FileReadMode:
-        return FileReadMode.READ
+        return FileReadMode.READ_BINARY
```
Comment (aldogonzalez8) [inline airbyte-cdk/python/airbyte_cdk/sources/file_based/file_types/blob_transfer.py:35]: I will remove this, you can ignore.

File: airbyte-cdk/python/airbyte_cdk/sources/file_based/file_types/file_transfer.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+    def get_file(
+        self,
+        config: FileBasedStreamConfig,
+        file: RemoteFile,
+        stream_reader: AbstractFileBasedStreamReader,
+        logger: logging.Logger,
+    ) -> Iterable[Dict[str, Any]]:
+        try:
+            yield stream_reader.get_file(file=file, local_directory=self._local_directory, logger=logger)
```
Comment (aldogonzalez8) [inline airbyte-cdk/python/airbyte_cdk/sources/file_based/file_types/file_transfer.py:28]: Implementation sample for source-s3 [here](https://github.com/airbytehq/airbyte/pull/48346/files#diff-b5e8d7cf7e79880dd2e77408410825335670decbc45f4960d2cf07ac323bae86).

Comment (aldogonzalez8) [reply airbyte-cdk/python/airbyte_cdk/sources/file_based/file_types/file_transfer.py:28]: Manual testing in [this workspace](https://cloud.airbyte.com/workspaces/166f3e04-9021-4f63-a963-a43abb548cec/connections/d7e111cf-1ddc-41bc-8b1a-34bf435c3f2f/timeline) seems be working fine for sync > 5GB :)

**Review by aaronsteers — APPROVED** [review]
I did a readthrough of all code changed. No issues found - looks good!

**Pull request merged**